connmgr/seed.go

```diff
@@ -25,60 +24,135 @@ const (
+// an HTTPS seeder.  It can be configured via the various exported functions
+// that start with the prefix SeedFilter.
+type HttpsSeederFilters struct {
+	ipVersion    uint16
+	hasIPVersion bool
+	pver         uint32
+	hasPver      bool
+	services     wire.ServiceFlag
+	hasServices  bool
+}
+
+// SeedFilterIPVersion configures a request to an HTTPS seeder to filter all
+// results that are not the provided ip version, which is expected to be either
+// 4 or 6, to indicate IPv4 or IPv6 addresses, respectively.  The HTTPS seeder
+// may choose to ignore other IP versions.
+func SeedFilterIPVersion(ipVersion uint16) func(f *HttpsSeederFilters) {
+	return func(f *HttpsSeederFilters) {
+		f.ipVersion = ipVersion
+		f.hasIPVersion = true
+	}
+}
+
+// SeedFilterProtocolVersion configures a request to an HTTPS seeder to filter
+// all results that are not the provided peer-to-peer protocol version.  This
+// can be useful to discover peers that support specific peer versions.
+func SeedFilterProtocolVersion(pver uint32) func(f *HttpsSeederFilters) {
+	return func(f *HttpsSeederFilters) {
+		f.pver = pver
+		f.hasPver = true
+	}
+}
+
+// SeedFilterServices configures a request to an HTTPS seeder to filter all
+// results that do not support the provided service flags.  This can be useful
+// to discover peers that support specific services such as fully-validating
+// nodes.
+func SeedFilterServices(services wire.ServiceFlag) func(f *HttpsSeederFilters) {
+	return func(f *HttpsSeederFilters) {
+		f.services = services
+		f.hasServices = true
+	}
+}
+
 // node defines a single JSON object returned by the https seeders.
 type node struct {
 	Host            string `json:"host"`
 	Services        uint64 `json:"services"`
 	ProtocolVersion uint32 `json:"pver"`
 }
 
-// SeedAddrs uses HTTPS seeding to populate the address manager with peers.
-func SeedAddrs(ctx context.Context, seeder string, ipversion uint16, pver uint32, services wire.ServiceFlag, dialFn DialFunc, seedFn OnSeed) error {
+// SeedAddrs uses HTTPS seeding to return a list of addresses of p2p peers on
+// the network.
+//
+// The seeder parameter specifies the domain name of the seed.  The dial
+// function specifies the dialer to use to contact the HTTPS seeder and allows
+// the caller to use whatever configuration it deems fit such as using a proxy,
+// like Tor.
+//
+// The variadic filters parameter allows additional query filters to be applied.
+// The available filters can be set via the exported functions that start with
+// the prefix SeedFilter.  See the documentation for each function for more
+// details.
+func SeedAddrs(ctx context.Context, seeder string, dialFn DialFunc, filters ...func(f *HttpsSeederFilters)) ([]*wire.NetAddress, error) {
+	// Set any caller provided filters.
+	var seederFilters HttpsSeederFilters
+	for _, f := range filters {
+		f(&seederFilters)
+	}
+
+	// Setup the HTTPS request.
 	req, err := http.NewRequest(http.MethodGet, "https://"+seeder, nil)
```

Comment:
I think this should be `NewRequestWithContext` so that the provided `ctx` controls execution of the request.

Reply:
The context is added just below via the `req = req.WithContext(ctx)` line, but I don't mind changing this instead.

Reply:
NewRequestWithContext was new to 1.13 iirc, hence the old code using the WithContext method.

Reply:
Updated to use `NewRequestWithContext` instead and removed the `req = req.WithContext(ctx)` line.

Reply:
I also bumped the `go.mod` to require `go 1.13` accordingly.